File: 동적계획법/leetcode53(Easy)-최대 서브 배열.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        sums = [nums[0]]
        for i in range(1, len(nums)):
            value = max(0, sums[i - 1]) + nums[i]
            sums.append(value)
        return max(sums)

# 카데인 알고리즘
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        best_sum = -sys.maxsize
        current_sum = 0
        
        for num in nums:
            current_sum = max(num, current_sum + num)
            best_sum = max(best_sum, current_sum)
        
        return best_sum

File: 동적계획법/leetcode53(Easy)-최대 서브 배열.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  const result = [nums[0]];

  for (let i = 1; i < nums.length; i += 1) {
    const value = Math.max(0, result[i - 1]) + nums[i];
    result.push(value);
  }

  return Math.max(...result);
};

// 카데인 알고리즘
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function (nums) {
  let bestSum = -Infinity;
  let currentSum = 0;

  nums.forEach((num) => {
    currentSum = Math.max(num, currentSum + num);
    bestSum = Math.max(currentSum, bestSum);
  });

  return bestSum;
};
